File: src/services/PDFGenerator.js
import { jsPDF } from 'jspdf';
import QRCode from 'qrcode';

// Configurações de tamanho e posições (em milímetros - mm)
const A4_WIDTH = 210;
const A4_HEIGHT = 297;

const drawAnchor = (doc, x, y, size = 10) => {
    // jsPDF rect: x, y, w, h, style ('F' = fill, 'D' = draw/stroke, 'DF' = draw and fill)
    doc.setFillColor(0, 0, 0); // Preto
    doc.rect(x - size / 2, y - size / 2, size, size, 'F');
};

const drawQRCode = async (doc, data, x, y, size = 30) => {
    try {
        // Gera o Data URI base64 da imagem do QR Code
        const qrDataUrl = await QRCode.toDataURL(data, {
            errorCorrectionLevel: 'L',
            margin: 0,
            width: size * 4 // Aumenta resolução do gerador para melhor qualidade no PDF
        });

        doc.addImage(qrDataUrl, 'PNG', x, y, size, size);
    } catch (error) {
        console.error("Erro ao gerar QR Code:", error);
    }
};

const drawSingleGabarito = async (doc, offsetX, offsetY, turmaNome, nomeAluno, alunoId, gabaritoId, numQuestoes, scale) => {
    // jsPDF não tem "c.saveState() / c.translate / c.scale" nativo como o ReportLab.
    // Portanto, criaremos uma função auxiliar que aplica as transformadas a todos os x, y desenhados.
    const transformX = (x) => offsetX + (x * scale);
    const transformY = (y) => offsetY + (y * scale);
    const transformSize = (s) => s * scale;

    const w = A4_WIDTH;
    const h = A4_HEIGHT;

    // Cabeçalho
    doc.setFont("helvetica", "bold");
    doc.setFontSize(16 * scale);
    doc.text("GABARITO FÁCIL - Sistema OMR", transformX(w / 2), transformY(25), { align: "center" });

    doc.setFont("helvetica", "normal");
    doc.setFontSize(12 * scale);
    doc.text(`Turma: ${turmaNome}`, transformX(w / 2), transformY(33), { align: "center" });

    if (nomeAluno) {
        doc.text(`Aluno: ${nomeAluno} (ID=${alunoId})`, transformX(w / 2), transformY(39), { align: "center" });
    }

    const qrData = JSON.stringify({ gabarito_id: gabaritoId, aluno_id: alunoId });

    // Âncoras OMR
    const margin = 10;
    const anchorSize = 10;

    // Top-Left (y = margin)
    drawAnchor(doc, transformX(margin + anchorSize / 2), transformY(margin + anchorSize / 2), transformSize(anchorSize));
    // Top-Right
    drawAnchor(doc, transformX(w - (margin + anchorSize / 2)), transformY(margin + anchorSize / 2), transformSize(anchorSize));
    // Bottom-Left (y = h - margin)
    drawAnchor(doc, transformX(margin + anchorSize / 2), transformY(h - (margin + anchorSize / 2)), transformSize(anchorSize));
    // Bottom-Right
    drawAnchor(doc, transformX(w - (margin + anchorSize / 2)), transformY(h - (margin + anchorSize / 2)), transformSize(anchorSize));

    // QR Code posicionado na parte inferior central
    await drawQRCode(doc, qrData, transformX((w - 25) / 2), transformY(h - margin - 25), transformSize(25));

    // Desenhar Bolinhas (O ReportLab a posição Y=0 é BOTTOM. No jsPDF a posição Y=0 é o TOP!)
    // Precisamos inverter a lógica de coordenadas Y das bolinhas.
    doc.setFont("helvetica", "bold");
    doc.setFontSize(10 * scale);

    const startX = 25;
    const startY = 75; // No python era `h - 75` vindo de baixo. Aqui é 75 vindo de cima.

    const qMarginY = 7;
    const qMarginX = 40;
    const altRadius = 2;

    const opcoes = ['A', 'B', 'C', 'D', 'E'];

    for (let q = 1; q <= numQuestoes; q++) {
        const col = Math.floor((q - 1) / 20);
        const row = (q - 1) % 20;

        const yPos = startY + (row * qMarginY); // Somar aumenta o Y (desce) no jsPDF
        const xPos = startX + (col * qMarginX);

        // Número da questão
        doc.setTextColor(0, 0, 0);
        doc.text(`${q.toString().padStart(2, '0')}.`, transformX(xPos), transformY(yPos + (altRadius / 2)));

        // Bolinhas
        doc.setLineWidth(0.5 * scale);
        for (let idx = 0; idx < opcoes.length; idx++) {
            const op = opcoes[idx];
            const bx = xPos + 10 + (idx * 6);
            const by = yPos;

            // Desenha Círculo
            doc.circle(transformX(bx), transformY(by), transformSize(altRadius), 'S'); // 'S' = Stroke

            // Letra de Dentro
            doc.setFont("helvetica", "normal");
            doc.setFontSize(6 * scale);
            doc.text(op, transformX(bx), transformY(by + 1), { align: "center" });

            // Retorna a fonte para negrito pro próximo número
            doc.setFont("helvetica", "bold");
            doc.setFontSize(10 * scale);
        }
    }
};

export const PDFGenerator = {
    async buildGabaritoPDF(turmaNome, alunos, gabaritoId, numQuestoes = 40, layout = 1) {
        const orientation = layout === 2 ? 'landscape' : 'portrait';
        // orientation: portrait/landscape, unit: mm, format: a4
        const doc = new jsPDF({
            orientation: orientation,
            unit: 'mm',
            format: 'a4'
        });

        // Se a lista estiver vazia (Gabarito Avulso), geramos a quantidade baseada no layout escolhido
        if (!alunos || alunos.length === 0) {
            alunos = Array(layout).fill({ id: "", nome: "" });
        }

        const wPage = layout === 2 ? A4_HEIGHT : A4_WIDTH;
        const hPage = layout === 2 ? A4_WIDTH : A4_HEIGHT;

        for (let i = 0; i < alunos.length; i += layout) {
            const lote = alunos.slice(i, i + layout);

            // Adiciona nova página a partir da segunda
            if (i > 0) {
                doc.addPage('a4', orientation);
            }

            if (layout === 1) {
                const aluno = lote[0];
                await drawSingleGabarito(doc, 0, 0, turmaNome, aluno.nome, aluno.id, gabaritoId, numQuestoes, 1.0);
            } else if (layout === 2) {
                const scale = 0.70;
                // As dimensões lógicas do gabarito (baseadas no A4 Portrait original)
                const logicalH = A4_HEIGHT * scale;
                const logicalW = A4_WIDTH * scale;

                // Em paisagem (w=297, h=210), alinhamos lado a lado (esquerda e direita)
                const offsetLeft = (wPage / 2 - logicalW) / 2;
                const offsetRight = (wPage / 2) + (wPage / 2 - logicalW) / 2;

                // Centraliza no eixo Y para não cortar
                const offsetY = (hPage - logicalH) / 2;

                // Linha de corte VERTICAL no meio da folha paisagem
                doc.setLineDashPattern([2, 1], 0);
                doc.setLineWidth(0.5);
                doc.line(wPage / 2, 0, wPage / 2, hPage);
                doc.setLineDashPattern([], 0); // Reset

                if (lote.length > 0) {
                    await drawSingleGabarito(doc, offsetLeft, offsetY, turmaNome, lote[0].nome, lote[0].id, gabaritoId, numQuestoes, scale);
                }
                if (lote.length > 1) {
                    await drawSingleGabarito(doc, offsetRight, offsetY, turmaNome, lote[1].nome, lote[1].id, gabaritoId, numQuestoes, scale);
                }
            } else if (layout === 4) {
                const scale = 0.48;
                const logicalH = hPage * scale;
                const logicalW = wPage * scale;

                // Linhas de corte em cruz
                doc.setLineDashPattern([2, 1], 0);
                doc.setLineWidth(0.5);
                doc.line(0, hPage / 2, wPage, hPage / 2);
                doc.line(wPage / 2, 0, wPage / 2, hPage);
                doc.setLineDashPattern([], 0);

                const gapX = (wPage / 2 - logicalW) / 2;
                const gapY = (hPage / 2 - logicalH) / 2;

                const positions = [
                    { x: gapX, y: gapY },                                      // Top-Left
                    { x: wPage / 2 + gapX, y: gapY },                          // Top-Right
                    { x: gapX, y: hPage / 2 + gapY },                          // Bottom-Left
                    { x: wPage / 2 + gapX, y: hPage / 2 + gapY }               // Bottom-Right
                ];

                for (let j = 0; j < lote.length; j++) {
                    const pos = positions[j];
                    await drawSingleGabarito(doc, pos.x, pos.y, turmaNome, lote[j].nome, lote[j].id, gabaritoId, numQuestoes, scale);
                }
            }
        }

        return doc;
    }
};

File: backend/services/pdf_generator.py
import io
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import mm
import qrcode

def mm_to_pt(mm_value):
    return mm_value * 2.83465

def draw_anchor(c, x, y, size=10*mm):
    # Desenha um quadrado escuro com borda (Âncora de Alinhamento OMR)
    c.setFillColorRGB(0, 0, 0)
    c.rect(x - size/2, y - size/2, size, size, fill=1)
    
def draw_qrcode(c, data, x, y, size=30*mm):
    qr = qrcode.QRCode(version=1, box_size=10, border=0)
    qr.add_data(data)
    qr.make(fit=True)
    img = qr.make_image(fill_color="black", back_color="white")
    
    # Salvar temporariamente na memoria
    import tempfile
    with tempfile.NamedTemporaryFile(suffix=".png") as temp_img:
        img.save(temp_img.name)
        c.drawImage(temp_img.name, x, y, width=size, height=size)

def draw_single_gabarito(c, offset_x, offset_y, turma_nome, nome_aluno, aluno_id, gabarito_id, num_questoes, scale):
    c.saveState()
    
    # 1. Translate point of origin to the block
    c.translate(offset_x, offset_y)
    
    # 2. Scale EVERYTHING inside this block to fit perfectly
    c.scale(scale, scale)
    
    # After scale, the logical width and height are still A4!
    w, h = A4
    
    c.setFont("Helvetica-Bold", 16)
    c.drawCentredString(w/2, h - 25*mm, "GABARITO FÁCIL - Sistema OMR")
    
    c.setFont("Helvetica", 12)
    c.drawCentredString(w/2, h - 33*mm, f"Turma: {turma_nome}")
    if nome_aluno:
        c.drawCentredString(w/2, h - 39*mm, f"Aluno: {nome_aluno} (ID={aluno_id})")
    
    qr_data = f'{{"gabarito_id":"{gabarito_id}","aluno_id":"{aluno_id}"}}'
    
    # Âncoras OMR para a Perspectiva do OpenCV
    margin = 10*mm # Reduzido para ficar mais na borda
    anchor_size = 10*mm
    draw_anchor(c, margin + anchor_size/2, h - (margin + anchor_size/2), anchor_size) # Top-Left
    draw_anchor(c, w - (margin + anchor_size/2), h - (margin + anchor_size/2), anchor_size) # Top-Right
    draw_anchor(c, margin + anchor_size/2, margin + anchor_size/2, anchor_size) # Bottom-Left
    draw_anchor(c, w - (margin + anchor_size/2), margin + anchor_size/2, anchor_size) # Bottom-Right
    
    # QR Code posicionado na parte inferior central
    draw_qrcode(c, qr_data, (w - 25*mm) / 2, margin, size=25*mm)
    
    # Desenhar as Bolinhas do Gabarito (Matriz)
    c.setFont("Helvetica-Bold", 10)
    start_x = 25*mm
    start_y = h - 75*mm # Aumentado para 75mm (antes 60mm) para mais espaço de cabeçalho
    
    q_margin_y = 7*mm
    q_margin_x = 40*mm
    alt_radius = 2*mm # Tamanho da bolinha
    
    opcoes = ['A', 'B', 'C', 'D', 'E']
    
    for q in range(1, num_questoes + 1):
        col = (q - 1) // 20
        row = (q - 1) % 20
        
        y_pos = start_y - (row * q_margin_y)
        x_pos = start_x + (col * q_margin_x)
        
        # Desenha número da questão
        c.setFillColorRGB(0, 0, 0)
        c.drawString(x_pos, y_pos - alt_radius/2, f"{q:02d}.")
        
        # Desenha bolinhas
        for idx, op in enumerate(opcoes):
            bx = x_pos + 10*mm + (idx * 6*mm)
            by = y_pos
            c.setLineWidth(0.5)
            c.circle(bx, by, alt_radius, stroke=1, fill=0)
            
            # Texto da alternativa (tamanho e opacidade baixos para não travar a câmera)
            c.setFont("Helvetica", 6)
            c.drawCentredString(bx, by - 1*mm, op)
            c.setFont("Helvetica-Bold", 10)

    # Restaurar estado (p/ voltar coordenada raiz 0,0)
    c.restoreState()


def build_gabarito_pdf(turma_nome: str, alunos: list, gabarito_id: str, num_questoes: int = 40, layout: int = 1):
    buffer = io.BytesIO()
    c = canvas.Canvas(buffer, pagesize=A4)
    w_page, h_page = A4

    # Se a lista estiver vazia, geramos 1 gabarito em branco
    if not alunos:
        alunos = [{"id": "", "nome": ""}]
        
    # Dividir alunos em lotes do tamanho do layout escolhido
    for i in range(0, len(alunos), layout):
        lote = alunos[i:i + layout]
        
        if layout == 1:
            aluno = lote[0]
            draw_single_gabarito(c, 0, 0, turma_nome, aluno["nome"], aluno["id"], gabarito_id, num_questoes, 1.0)
            
        elif layout == 2:
            scale = 0.70 
            logical_h = h_page * scale
            logical_w = w_page * scale
            offset_x = (w_page - logical_w) / 2
            
            offset_y_top = h_page/2 + (h_page/2 - logical_h) / 2
            offset_y_bottom = (h_page/2 - logical_h) / 2
            
            c.setDash(6, 3)
            c.setLineWidth(0.5)
            c.line(0, h_page/2, w_page, h_page/2)
            c.setDash()
            
            # Top
            if len(lote) > 0:
                draw_single_gabarito(c, offset_x, offset_y_top, turma_nome, lote[0]["nome"], lote[0]["id"], gabarito_id, num_questoes, scale)
            # Bottom
            if len(lote) > 1:
                draw_single_gabarito(c, offset_x, offset_y_bottom, turma_nome, lote[1]["nome"], lote[1]["id"], gabarito_id, num_questoes, scale)
                
        elif layout == 4:
            scale = 0.48
            logical_h = h_page * scale
            logical_w = w_page * scale
            
            c.setDash(6, 3)
            c.setLineWidth(0.5)
            c.line(0, h_page/2, w_page, h_page/2)
            c.line(w_page/2, 0, w_page/2, h_page)
            c.setDash()
            
            gap_x = (w_page/2 - logical_w) / 2
            gap_y = (h_page/2 - logical_h) / 2
            
            positions = [
                (gap_x, h_page/2 + gap_y),          # Top-Left
                (w_page/2 + gap_x, h_page/2 + gap_y), # Top-Right
                (gap_x, gap_y),                     # Bottom-Left
                (w_page/2 + gap_x, gap_y)            # Bottom-Right
            ]
            
            for j, aluno_data in enumerate(lote):
                pos_x, pos_y = positions[j]
                draw_single_gabarito(c, pos_x, pos_y, turma_nome, aluno_data["nome"], aluno_data["id"], gabarito_id, num_questoes, scale)

        # Após desenhar os gabaritos dessa folha, vira a página
        c.showPage()
        
    c.save()
    
    buffer.seek(0)
    return buffer

